### ISSUE TITLE
Using embind to generate TypeScript typings

### ISSUE THREAD

**Issue by boyanio** [emscripten-core/emscripten#7083]
Hello. To export C++ structs and classes, one can use embind. As the export is very explicit, I was thinking, one could use embind to generate TypeScript typings (.d.ts) too. I would be happy to hear your comments on that :-)

**Comment by isc30** [comment]
Yes, I like that idea. I would go even further and introduce some kind of "ownership" expressed by marker interfaces. Something like this:

    interface owns<T> extends T
    {
        delete(): void;
    }

This way your API can be explicit on what objects you need to call `.delete()` (this is very similar to IDisposable interface in C#).

**Comment by boyanio** [comment]
Good point! If we look at the [class example in the documentation](https://kripken.github.io/emscripten-site/docs/porting/connecting_cpp_and_javascript/embind.html#classes), we could generate the following typings:

```
declare namespace Module {
    export interface Deletable {
        delete();
    }

    export interface MyClass extends Deletable {
        x: number;
        incrementX(): void;
    }

    export var MyClass: {
        new(x: number, y: string): MyClass;

        getStringFromInstance(a: MyClass): string;
    }
}
```

As not everyone uses TypeScript, this part should be optional. Could be a separate bind switch, for example

`emcc --bind --ts-typings -o quick_example.js quick_example.cpp`

**Comment by isc30** [comment]
I think the interface inheritance should go the other way around:

```ts
declare namespace Module {
    export interface Owner<T> : T {
        delete();
    }

    export interface MyClass {
        x: number;
        incrementX(): void;
    }

    export var MyClass: {
        new(x: number, y: string): Owner<MyClass>; // js needs to destroy it
        getSomeNotOwningPointer(): MyClass; // js doesnt need to destroy it

        getStringFromInstance(a: MyClass): string;
    }
}
```

**Comment by k2snowman69** [comment]
Has anyone here had success properly typing the output library? For example, given the EMBIND syntax 
```
EMSCRIPTEN_BINDINGS(StringFactory) {
    class_<StringFactory>("StringFactory")
        .smart_ptr_constructor("StringFactory", &std::make_shared<StringFactory>)
        .function("getString", &StringFactory::getString)
        ;
}
```
And your make command uses the `-s Modularize` which means the consumption of the library would be as follows:
```
import Module from 'string-factory';

instance = Module({
    onRuntimeInitialized: () => {
        let stringFactory = new instance.StringFactory();
        stringFactory.getString();
    },
});
```

The issue here is that Module isn't exporting as the default export, it's exporting as all exports. This makes writing a single d.ts file difficult

```
declare namespace ClientShared {
    export interface StringFactory {
        new(): StringFactory;
        getString(): string;
    }
    export interface Lib {
        StringFactory: StringFactory;
    }
}
declare function Module (emscriptenArgs: any): ClientShared.Lib;
export = Module;
```

Because in the code you can't seem to reference `ClientShared.StringFactory` as a type. 

So would this mean that emscripten would have to output as a module vs a library (e.g. `export default Module` vs `export = Module`)? Or did I miss something in the typescript documentation?

**Comment by k2snowman69** [comment]
To answer my own question, to be able to use a d.ts similar to what I last described you'll need to add `-s EXPORT_ES6` to the command line. Further I created https://github.com/kripken/emscripten/pull/7207 which will make the exported library written in ES5 but consumable via the `import Module, {<insert types here>} from …` syntax without requiring a transpile before consumption.

**Comment by endel** [comment]
I believe this is totally possible, given the verbosity and details we provide to `EMSCRIPTEN_BINDINGS`. Has anyone started implementing this? I'd love to contribute!

**Comment by isc30** [comment]
as far as I know no one started working on this

**Comment by wffurr** [comment]
I'm working with a prototype of this, created by a colleague.  If I can get it working on a couple of different embind examples, then I will see about contributing it upstream.

**Comment by mathiasi** [comment]
Any updates on this?

**Comment by wffurr** [comment]
The tool itself works fine for the subset of embind it supports.  I'm having trouble making it work for my project due to project-specific issues.  I'm also not sure the best way to integrate it with the Emscripten toolchain.  I might have something by the end of the year.

**Comment by kripken** [comment]
Reopening, as this still sounds useful. @wffurr , any updates?

**Comment by netpoetica** [comment]
@jjrv Did a really great job on this with `nbind`. I would if they might be able to provide some insight to help bring this directly into emscripten? Or if someone who is more familiar with emscripten may be able to draw some inspiration from their work: https://github.com/charto/nbind/blob/master/bin/ndts.js

I have been working with WebIDL with emscripten and auto-generating types from it, but I have run into a brick wall because WebIDL is incredibly unsophisticated when it comes to generic, templates and typedefs, which is one of the main things you want from C/C++ world. I think at this point for sure I will need to move over to using `embind` api and manually writing type definitions. I am not sure I am savvy enough on the C++ or emscripten side to contribute auto-generated types but I sure do wish I was :)

**Comment by netpoetica** [comment]
@wffurr Some of the code in `nbind` may provide some useful insight for you if you're still interested in tackling this: https://github.com/charto/nbind/blob/master/src/todts.ts and the Reflect.cc/Reflect.ts files I would imagine are the main thing

**Comment by ted537** [comment]
https://github.com/ted537/tsembind

I've created a rudimentary typescript bindings generator that works on existing EMBind libraries. There are some hacky things present (mostly grabbing context) but are only necessary for an external implementation.

**Comment by shivdeepgarud** [comment]
@ted537 Am I correct in understanding that your generator will take the abc.js file generated by emscripten during compilation and output a ts file? I tried installing it from your github link but all it installs are a bunch of js/ts files, no executable. Anything I am missing? By the way, we were thinking of writing a js script that would read c++ source files and generate a module.d.ts file. I think your project already does that?

**Comment by ssg3d** [comment]
@ted537 Am I correct in understanding that your generator will take the abc.js file generated by emscripten during compilation and output a ts file? I tried installing it from your github link but all it installs are a bunch of js/ts files, no executable. Anything I am missing? By the way, we were thinking of writing a js script that would read c++ source files and generate a module.d.ts file. I think your project already does that?

**Comment by ssg3d** [comment]
Thanks! I get this error when I try to run this on my js file though -

failed to asynchronously prepare wasm: LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object
LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object
RuntimeError: abort(LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object) at Error

**Comment by ted537** [comment]
> Thanks! I get this error when I try to run this on my js file though -
> 
> failed to asynchronously prepare wasm: LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object
> LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object
> RuntimeError: abort(LinkError: WebAssembly.instantiate(): Import #114 module="env" function="memory" error: memory import must be a WebAssembly.Memory object) at Error

Could you create an [issue](https://github.com/ted537/tsembind/issues) for this? I've only tested it with some of the formats for the `.js` file Emscripten generates.

**Comment by ssg3d** [comment]
Done!

**Comment by mmarczell-graphisoft** [comment]
@ted537 
> A similar strategy should be possible from the C++ side as well. I imagine a drop in `bind-with-types.h` header (which wraps `emscripten/bind.h` functionality) would be sufficient.

I am creating this right now in our product. The approach I took is to compile all the binding C++ code twice, with the second pass using a replacement header and library that replicates Embind's API (instead of the wrapping you mentioned here). Running this second executable writes the declarations to a file and this can be nicely automated from CMake as a custom build step.

**Comment by devshgraphicsprogramming** [comment]
> @ted537
> 
> > A similar strategy should be possible from the C++ side as well. I imagine a drop in `bind-with-types.h` header (which wraps `emscripten/bind.h` functionality) would be sufficient.
> 
> I am creating this right now in our product. The approach I took is to compile all the binding C++ code twice, with the second pass using a replacement header and library that replicates Embind's API (instead of the wrapping you mentioned here). Running this second executable writes the declarations to a file and this can be nicely automated from CMake as a custom build step.

lovely, can you make the repo public?

**Comment by devshgraphicsprogramming** [comment]
> > > @ted537
> > > lovely, can you make the repo public so we could contribute?
> 
> I'd love to! But it's not a "repo" I could just "make public" - we use Perforce, and this code is not an isolated library or tool, it is part of our larger codebase. I'd have to work to isolate it, write a readme, etc. and I'd have to go through some corporate procedure to open source it.

Well, if you go through all the pain, the least I can promise you that we'll try to use it in production and contribute a bit.

**Comment by gracicot** [comment]
@mmarczell-graphisoft Honestly, even a simple gists of that replacement header file would be of extreme help. I would contribut too to make it production ready

**Comment by devshgraphicsprogramming** [comment]
We actually submitted a talk to WASM I/O about generating JavaScript bindings to C# with C++ Reflection TS, Cross Language Polymorphism, Embind and SWIG

Unfortunately got rejected, barrier to entry was too high XD

**Comment by marczellm** [comment]
@gracicot @devshgraphicsprogramming 
Here you are: https://github.com/marczellm/emscripdtsgen
However I can see that https://github.com/ted537/tsembind is seeing much more active development.

**Comment by devshgraphicsprogramming** [comment]
> @gracicot @devshgraphicsprogramming Here you are: https://github.com/marczellm/emscripdtsgen However I can see that https://github.com/ted537/tsembind is seeing much more active development.

TSembind completely failed us with regards to namespaces, nested classes, etc.

I'm hoping that our client allows us to finally mention them by name and accepts our proposal to let us speak about the tooling we developed at WASM Summit.

**Comment by mike-lischke** [comment]
@marczellm tsembind is useless for ES6 modules, it only works with CommonJS output.

**Comment by devshgraphicsprogramming** [comment]
> @marczellm tsembind is useless for ES6 modules, it only works with CommonJS output.

moreover TSembind won't reconstruct your namespaces and nested classes.

**Comment by brendandahl** [comment]
I've landed the new emcc option `--embind-emit-tsd <filename>`. When using the flag your program will be compiled and run under node to generate the TS definitions. The feature is still new, but is supports most of the commonly used embind features.  Please try it out and file new issues for feature requests or bugs.

Closing at fixed.

**Comment by mike-lischke** [comment]
@brendandahl Is this already published? I upgraded to emscripten 3.1.43 and added that option to my build script, but instead of a file I got a number of warnings (_embind_register_class_class_property) and a final error 

> <path>/emscripten_zdgc_erx.rsp.utf-8' failed (returned 1)

If it is already released I can open a new issue for this.

**Comment by starball5** [comment]
Thanks for the work being done here! I've written up a TL;DR on Stack Overflow [here](https://stackoverflow.com/q/76691590/11107541) that I'll keep updated as things progress.
